analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/Extensions/BinaryOperatorKindExtensions.cs

```diff
@@ -47,4 +49,56 @@ kind switch
             BinaryOperatorKind.LessThanOrEqual when left.Min > right.Max => BoolConstraint.False,
             _ => null
         };
+
+    public static NumberConstraint NumberConstraintFromRelationalOperator(this BinaryOperatorKind kind, NumberConstraint existingNumber, NumberConstraint comparedNumber) =>
+        kind switch
+        {
+            BinaryOperatorKind.Equals => NumberConstraint.From(
+                ArithmeticCalculator.BiggestMinimum(comparedNumber, existingNumber),
+                ArithmeticCalculator.SmallestMaximum(comparedNumber, existingNumber)),
+            BinaryOperatorKind.NotEquals when comparedNumber.IsSingleValue && comparedNumber.Min == existingNumber?.Min =>
+                NumberConstraint.From(existingNumber!.Min + 1, existingNumber.Max),
+            BinaryOperatorKind.NotEquals when comparedNumber.IsSingleValue && comparedNumber.Min == existingNumber?.Max =>
+                NumberConstraint.From(existingNumber!.Min, existingNumber.Max - 1),
+            BinaryOperatorKind.GreaterThan when comparedNumber.Min.HasValue =>
+                From(comparedNumber.Min + 1, null, existingNumber),
+            BinaryOperatorKind.GreaterThanOrEqual when comparedNumber.Min.HasValue =>
+                From(comparedNumber.Min, null, existingNumber),
+            BinaryOperatorKind.LessThan when comparedNumber.Max.HasValue =>
+                From(null, comparedNumber.Max - 1, existingNumber),
+            BinaryOperatorKind.LessThanOrEqual when comparedNumber.Max.HasValue =>
+                From(null, comparedNumber.Max, existingNumber),
+            _ => null
+        };
+
+    private static NumberConstraint From(BigInteger? newMin, BigInteger? newMax, NumberConstraint existingNumber)
+    {
+        if (existingNumber is not null)
+        {
+            if (newMin is null || existingNumber.Min > newMin)
+            {
+                newMin = existingNumber.Min;
+            }
+            if (newMax is null || existingNumber.Max < newMax)
+            {
+                newMax = existingNumber.Max;
+            }
+        }
+        return NumberConstraint.From(newMin, newMax);
+    }
+
+    public static BinaryOperatorKind ApplyOpposite(this BinaryOperatorKind kind, bool useOpposite) =>
```

Comment:
How about we call both of the methods `Opposite`?

Reply:
I was thinking that, but I followed the naming scheme of `SymbolicConstraint`.